---
title: "R Notebook"
output: html_notebook
---

```{r}
library(Biostrings)
library(rtracklayer)
library(tidyverse)
```

## Obtain following files to estimate CAI and tAI values
```{r}
# Ribosomal genes for CAI
path <- "./"
ribo.genes <- read_csv(paste0(path, "human.ribo.csv")) %>% dplyr::rename(Gene = `Approved symbol`) 

### Read in codon-specific TAI weights values
tai.w <- read_delim(paste0(path, "tuller.taiw.tsv"), "\t")
tai.w <- tai.w %>% arrange(Codon) %>% mutate(w = as.numeric(w))
```

```{r}
# Files below were generated using script_for_transcript_annotation file
appris <- read_delim(paste0(path, "/human_appris_data.principal.txt"), "\t", col_names = F) %>% 
  dplyr::rename(Gene = X1, ENSEMBL = X2, Transcript = X3)
Hs_GRanges <- readGFFAsGRanges(paste0(path,"human_appris_cdna.gff3")) # Reading in GFF file with 5'UTR, CDS, 3'UTR annotations

Hs_cDNA <- readDNAStringSet(paste0(path,"human_appris_cdna.fa.gz"))  # Reading in nucleotide sequences for each transcript
```

```{r}
Hs_cDNA <- Hs_cDNA[order(names(Hs_cDNA))] # Ordering by Ensembl ID
Hs_GRanges <- Hs_GRanges[order(Hs_GRanges$Name),] # Ordering by Ensembl IDs

# Check if the order of transcripts in GRanges matches with DNAStringSet
sum(names(Hs_cDNA) != Hs_GRanges$Name[Hs_GRanges$type == "CDS"])
sum(names(Hs_cDNA) != Hs_GRanges$Name[Hs_GRanges$type == "UTR5"])
sum(names(Hs_cDNA) != Hs_GRanges$Name[Hs_GRanges$type == "UTR3"])
```

#### Formatting data into appropriate DNAStringSet and GRanges objects
```{r}
Hs_GRanges$Length <- width(Hs_GRanges)

final_table <- mcols(Hs_GRanges) %>% 
  as_tibble() %>%
  select(Name, type, Length) %>%
  spread(type, Length) %>%
  rename(Transcript = Name)

```

#### Formatting DNAStringSet by GRange Information
```{r}
# Collect the nucleotide sequences for the 5'UTR, CDS, 3'UTR of a given gene
Hs_cDNA_UTR5 <- subseq(Hs_cDNA, 
                       start = start(Hs_GRanges[Hs_GRanges$type == "UTR5"]),
                       end = end(Hs_GRanges[Hs_GRanges$type == "UTR5"]))
Hs_cDNA_CDS  <- subseq(Hs_cDNA,
                       start = start(Hs_GRanges[Hs_GRanges$type == "CDS"]),
                       end = end(Hs_GRanges[Hs_GRanges$type == "CDS"]))
Hs_cDNA_UTR3 <- subseq(Hs_cDNA,
                       start = start(Hs_GRanges[Hs_GRanges$type == "UTR3"]),
                       end = end(Hs_GRanges[Hs_GRanges$type == "UTR3"]))
```

#### Calculating 5'UTR GC Content, CDS GC Content, and upstream AUG (uAUGs) Counts
```{r}
# GC Content and uATG counts
UTR5_GC_Content <- tibble(Transcript = names(Hs_cDNA_UTR5),
                          UTR5_GC_Content = c(letterFrequency(Hs_cDNA_UTR5, "CG", as.prob = T)))
UTR3_GC_Content <- tibble(Transcript = names(Hs_cDNA_UTR3),
                          UTR3_GC_Content = c(letterFrequency(Hs_cDNA_UTR3, "CG", as.prob = T)))
CDS_GC_Content <- tibble(Transcript = names(Hs_cDNA_CDS),
                         CDS_GC_Content = c(letterFrequency(Hs_cDNA_CDS, "CG", as.prob = T)))
uATG_count <- tibble(Transcript = names(Hs_cDNA_UTR5),
                     uATG_count = c(oligonucleotideFrequency(Hs_cDNA_UTR5, 3)[, "ATG"]))

final_table <- plyr::join_all(list(final_table, UTR5_GC_Content, UTR3_GC_Content, CDS_GC_Content, uATG_count), 
                              by = "Transcript", 
                              type = "left") %>% as_tibble()
```


#### Formating data for RNAfold 
[RNAfold](https://www.tbi.univie.ac.at/RNA/RNAfold.1.html) determines the minimum free energy of secondary structures. We determined the free energy of two regions of a given gene: 1.) transcription start site and 2.) translation start site. For the transcription start site, we took the first 30 nts of the 5'UTR. For translation start sites, we took 10 nts upstream of the start codon and 20 nts downstream of the start codon.
```{r}
# Isolating the first 30 nts of 5'UTR to determine folding energy
Hs_cDNA_TranscriptionEnergy <- subseq(Hs_cDNA, 1, 30)

# Isolating the last 10 nts of 5'UTR and first 20 nts of the CDS to determine folding energy
Hs_cDNA_TranslationEnergy <- Hs_cDNA_UTR5

index_lessthan10nt <- width(Hs_cDNA_UTR5) < 10 # Identify genes with UTR5 seqs less than 10 nts
Hs_cDNA_TranslationEnergy[index_lessthan10nt] <- subseq(Hs_cDNA[index_lessthan10nt], 1, 30) # For these genes, calculate FE based on first 30nts of CDS

Hs_cDNA_UTR5_gt10 <- Hs_cDNA_UTR5[!index_lessthan10nt]
Hs_cDNA_CDS_gt10 <- Hs_cDNA_CDS[!index_lessthan10nt]

# Isolating the last 10 nts of 5'UTR and first 20 nts of the CDS to determine folding energy
Hs_cDNA_TranslationEnergy[!index_lessthan10nt] <- xscat(subseq(Hs_cDNA_UTR5_gt10,
                                                               start = (nchar(Hs_cDNA_UTR5_gt10) - 9),
                                                               end = nchar(Hs_cDNA_UTR5_gt10)),
                                                        subseq(Hs_cDNA_CDS_gt10, 1, 20))
```

```{r}
# Exporting FASTA file for RNAfold
output <- paste0(path,"RNAfold_Data/Hs_cDNA_TranscriptionEnergy.fa")
writeXStringSet(Hs_cDNA_TranscriptionEnergy, output)
 
output2 <- paste0(path,"RNAfold_Data/Hs_cDNA_TranslationEnergy.fa")
writeXStringSet(Hs_cDNA_TranslationEnergy, output2)
```

### Calculate folding energies using RNAfold
```{bash}
export PATH=${HOME}/ViennaRNA_Package/Program:${PATH}
RNAfold < ./RNAfold_Data/Hs_cDNA_TranscriptionEnergy.fa > ./RNAfold_Data/Hs_cDNA_TranscriptionEnergy.out
RNAfold < ./RNAfold_Data/Hs_cDNA_TranslationEnergy.fa > ./RNAfold_Data/Hs_cDNA_TranslationEnergy.out
```

The `R` code (written with the heading `{r}` above) is needed to export the RNAfold inputs as a FASTA file (required by RNAfold program). The `{bash}` read out above is the Unix commands needed to execute the RNAfold program on the FASTA inputs as well as define the RNAfold output files and their locations. The RNAfold was executive with all defaults. **Used ViennaRNA-2.3.5. Note: The "/ViennaRNA_Package/Program" in the {bash} is the folder that has the RNAfold Unix executable program.**

#### Accessing the RNAfold output
```{r, message=FALSE}
# RNAfold output
# Transcription Folding Energy
Hs_cDNA_TranscriptionEnergy_output <- read_csv(paste0(path,"RNAfold_Data/Hs_cDNA_TranscriptionEnergy.out"), col_names = FALSE)

# Translation Folding Energy
Hs_cDNA_TranslationEnergy_output <- read_csv(paste0(path,"RNAfold_Data/Hs_cDNA_TranslationEnergy.out"), col_names = FALSE)
```

#### Formatting the Output of RNAfold
```{r}
# Transcription Folding Energy
# Transcript name was every first cell and energy was every third
Transcript_FE <- tibble(
  Transcript = Hs_cDNA_TranscriptionEnergy_output$X1[seq(1, nrow(Hs_cDNA_TranscriptionEnergy_output), 3)],
  Transcript_FE = Hs_cDNA_TranscriptionEnergy_output$X1[seq(3, nrow(Hs_cDNA_TranscriptionEnergy_output), 3)]) %>%
  mutate(Transcript = substr(Transcript, 2, nchar(Transcript)),
         Transcript_FE = as.numeric(gsub("[\\(\\)]", "", sapply(strsplit(Transcript_FE," "), tail, 1))))

# Translation Folding Energy
Translate_FE <- tibble(
  Transcript = Hs_cDNA_TranslationEnergy_output$X1[seq(1, nrow(Hs_cDNA_TranslationEnergy_output),3)],
  Translate_FE = Hs_cDNA_TranslationEnergy_output$X1[seq(3, nrow(Hs_cDNA_TranslationEnergy_output),3)]) %>%
  mutate(Transcript = substr(Transcript, 2, nchar(Transcript)),
         Translate_FE = as.numeric(gsub("[\\(\\)]", "", sapply(strsplit(Translate_FE," "), tail, 1))))
```

#### Placing Folding Energies (FE) for genes in `final_table`
```{r}
final_table <- final_table %>%
  left_join(., Transcript_FE, "Transcript") %>%
  left_join(., Translate_FE, "Transcript")
```

### Calculate TAI
```{r}
codon.freq <- oligonucleotideFrequency(subseq(Hs_cDNA_CDS, 1, width(Hs_cDNA_CDS)-3), 3, step = 3)
codon.w <- t(t(codon.freq) * log(tai.w$w))
codon.w <- cbind(codon.w, rowSums(codon.freq))
tAI <- apply(codon.w, 1, function(x){exp(sum(x[1:64])/x[65])}) 

final_table <- final_table %>%
  left_join(., tibble(Transcript = names(Hs_cDNA_CDS), tAI = tAI), "Transcript")
```

### Calculate CAI
```{r}
ribo.genes <- ribo.genes %>% filter(grepl("^ribo", `Approved name`))

appris <- appris %>% filter(Transcript %in% final_table$Transcript)

ribo.transcript <- left_join(ribo.genes, appris, "Gene") %>% filter(!is.na(Transcript))
ribo.transcript.seq <- Hs_cDNA_CDS[names(Hs_cDNA_CDS) %in% ribo.transcript$Transcript]
```

```{r}
ribo.codon.freq <- oligonucleotideFrequency(subseq(ribo.transcript.seq, 1, width(ribo.transcript.seq)-3), 3, step = 3) %>% colSums()

ribo.w.table <- left_join(tai.w, tibble(Codon = names(ribo.codon.freq), Count = ribo.codon.freq), "Codon")
ribo.w.table <- ribo.w.table %>%
  group_by(AA) %>%
  dplyr::mutate(cai.w = Count/max(Count))

codon.w <- t(t(codon.freq) * log(ribo.w.table$cai.w))
codon.w <- cbind(codon.w, rowSums(codon.freq))
CAI <- apply(codon.w, 1, function(x){exp(sum(x[1:64], na.rm = T)/x[65])}) 

final_table <- final_table %>%
  left_join(., tibble(Transcript = names(Hs_cDNA_CDS), CAI = CAI), "Transcript")
```

```{r}
write_feather(final_table, paste0(path, "human_features.feather"))
```

