# References

Citations for third-party software used by RiboViz (in order of package name).

Quinlan, A.R. and Hall I.M. (2010) BEDTools: a flexible suite of utilities for comparing genomic features. Bioinformatics. 26, 6, pp. 841-842. doi: [10.1093/bioinformatics/btq033](https://doi.org/10.1093/bioinformatics/btq033).

Pagès, H., Aboyoun, P., Gentleman R. and DebRoy, S. (2017) Biostrings: Efficient manipulation of biological strings. R package version 2.46.0. doi: [10.18129/B9.bioc.Biostrings](https://doi.org/10.18129/B9.bioc.Biostrings).

Langmead, B. and Salzberg S. (2012) Fast gapped-read alignment with Bowtie 2. Nature Methods. 2012, 9:357-359. doi: [10.1038/nmeth.1923](https://doi.org/10.1038/nmeth.1923).

Cock, P.J.A., Antao, T., Chang, J.T., Chapman, B.A., Cox, C.J., Dalke, A., Friedberg, I., Hamelryck, T., Kauff, F., Wilczynski, B. and de Hoon, M.J.L. (2009) Biopython: freely available Python tools for computational molecular biology and bioinformatics, Bioinformatics, Volume 25, Issue 11, 1 June 2009, Pages 1422-1423, doi: [10.1093/bioinformatics/btp163](https://doi.org/10.1093/bioinformatics/btp163).

Martin, M. (2011) Cutadapt removes adapter sequences from high-throughput sequencing reads. EMBnet.journal, 17(1), pp. 10-12. doi:[10.14806/ej.17.1.200](https://doi.org/10.14806/ej.17.1.200).

gffutils, http://daler.github.io/gffutils/

Wickham, H. (2016) ggplot2: Elegant Graphics for Data Analysis. Springer-Verlag New York, 2016. ISBN 978-0-387-98141-3.

Widgren, S., et al. (2018) git2r: Provides Access to Git Repositories. R package version 0.21.0. URL https://CRAN.R-project.org/package=git2r.

gitpython, https://github.com/gitpython-developers/GitPython/

GraphViz, https://www.graphviz.org/

h5py, https://www.h5py.org/

The HDF Group. Hierarchical Data Format, version 5, 1997-2020, http://www.hdfgroup.org/HDF5/.

Kim D., Langmead, B. and Salzberg, S.L. (2015) HISAT: a fast spliced aligner with low memory requirements. Nature Methods 2015. doi: [10.1038/nmeth.3317](https://doi.org/10.1038/nmeth.3317).

Hester, J. and Angly, F. (2019) lintr: A 'Linter' for R Code. R package version 2.0.0. https://CRAN.R-project.org/package=lintr

Davis, T.L. (2019) optparse: Command Line Option Parser. R package version 1.6.4. https://CRAN.R-project.org/package=optparse.

McKinney, W. (2010) [Data Structures for Statistical Computing in Python](http://conference.scipy.org/proceedings/scipy2010/mckinney.html), Proceedings of the 9th Python in Science Conference, 51-56.

Reback, J., McKinney, W., jbrockmendel, den Bossche, J.V., Augspurger, T., Cloud, P. and Whelan, C. (2020) pandas-dev/pandas. Zenodo. doi: [10.5281/zenodo.3644238](http://doi.org/10.5281/zenodo.3644238).

pigz, http://zlib.net/pigz/

Sievert, S. (2018) plotly for R. https://plotly-r.com/

pycodestyle, https://github.com/pycqa/pycodestyle/

pylint, https://www.pylint.org/

Li et al. (2009) The Sequence Alignment/Map format and SAMtools, Bioinformatics. 2009 Aug 15;25(16):2078-9. doi:[10.1093/bioinformatics/btp352](https://doi.org/10.1093/bioinformatics/btp352)

pytest, https://github.com/pytest-dev/pytest/

pytest-cov, https://github.com/pytest-dev/pytest-cov/

pyyaml, https://pyyaml.org/

R Core Team (2020) R: A Language and Environment for Statistical Computing, R Foundation for Statistical Computing, Vienna, Austria, 2020. https://www.R-project.org.

Ushey, K. (2018) RcppRoll: Efficient Rolling / Windowed Operations. R package version 0.3.0. https://CRAN.R-project.org/package=RcppRoll.

Wickham, W., Hester, J. and Francois, R. (2018) readr: Read Rectangular Text Data. R package version 1.3.1. https://CRAN.R-project.org/package=readr.

Fischer, B., Pau, G. and Smith, M. (2017) rhdf5: HDF5 interface to R. R package version 2.22.0, https://github.com/grimbough/rhdf5.

Morgan, M., Pagès, J., Obenchain V. and Hayden, N. (2017) Rsamtools: Binary alignment (BAM), FASTA, variant call (BCF), and tabix file import. R package version 1.30.0. http://bioconductor.org/packages/release/bioc/html/Rsamtools.html.

Lawrence, M., Gentleman, R. and Carey, V. (2009) rtracklayer: an {R} package for interfacing with genome browsers. Bioinformatics 25:1841-1842. doi: [10.1093/bioinformatics/btp328](http://bioinformatics.oxfordjournals.org/content/25/14/1841.abstract).

Chang, W., Cheng, J., Allaire, J.J., Xie Y. and McPherson, J (2019) shiny: Web Application Framework for R. R package version 1.3.2. https://CRAN.R-project.org/package=shiny

Morgan, M., Anders, S., Lawrence, M., Aboyoun, P., Pagès, H. and Gentleman, R. (2009) ShortRead: a Bioconductor package for input, quality assessment and exploration of high-throughput sequence data. Bioinformatics 25:2607-2608. doi: [10.1093/bioinformatics/btp450](http://dx.doi.org/10.1093/bioinformatics/btp450).

Sphinx, https://www.sphinx-doc.org/

Müller, K. and Walthert, L. (2019) styler: Non-Invasive Pretty Printing of R Code. R package version 1.2.0. https://CRAN.R-project.org/package=styler.

Wickham, H. and Henry, L. (2019) tidyr: Tidy Messy Data. R package version 1.0.0. https://CRAN.R-project.org/package=tidyr.

Smith, T.S., Heger , A. and Sudbery, I. (2017) UMI-tools: Modelling sequencing errors in Unique Molecular Identifiers to improve quantification accuracy. Genome Res. 2017, January 18 2017, doi: [10.1101/gr.209601.116](https://doi.org/10.1101/gr.209601.116).
